force-app/main/default/classes/ServiceDeliveryService.cls

```diff
@@ -0,0 +1,201 @@
+/*
+ *
+ *  * Copyright (c) 2020, salesforce.com, inc.
+ *  * All rights reserved.
+ *  * SPDX-License-Identifier: BSD-3-Clause
+ *  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ *
+ */
+
+public with sharing class ServiceDeliveryService {
+    @TestVisible
+    private AttendanceSelector attendanceSelector = new AttendanceSelector();
+    @TestVisible
+    private ServiceDeliveryDomain domain = new ServiceDeliveryDomain();
+    private Rollups rollups;
+    public List<String> presentStatuses = attendanceSelector.presentStatuses;
+    public List<String> absentStatuses = attendanceSelector.absentStatuses;
+    private static final String DEFAULT_STATUS = 'Present';
+    // While we allow the Attendance UI and the ultimate saved Service Delivery records
+    // to be driven by a field set, the mapping from Service Participants to the seeded
+    // Service Deliveries that appear in the Attendance UI is hard-coded here.
+    // We could, in the future, allow this to be driven by design parameters or CMDT.
+    // It is a heavy lift for questionable value. (Is there a use case for mapping additional
+    // information from Participant to Attendance Service Delivery?)
+    private static final Map<SObjectField, SObjectField> SERVICE_FIELD_BY_PARTICIPANT_FIELD = new Map<SObjectField, SObjectField>{
+        ServiceParticipant__c.Contact__c => ServiceDelivery__c.Contact__c,
+        ServiceParticipant__c.ProgramEngagement__c => ServiceDelivery__c.ProgramEngagement__c
+    };
+
+    public List<ServiceDelivery__c> generateRoster(Id sessionId) {
+        return createRows(sessionId);
+    }
+
+    public void upsertServiceDeliveries(
+        List<ServiceDelivery__c> serviceDeliveriesToUpsert
+    ) {
+        Savepoint savepoint = Database.setSavepoint();
+        try {
+            domain.upsertServiceDeliveries(serviceDeliveriesToUpsert);
+        } catch (Exception ex) {
+            Database.rollback(savepoint);
+            throw new ServiceDeliveryException(ex);
+        }
+    }
+
+    public void doRollups(Map<SObjectField, Set<Id>> recordIdsByLookupField) {
+        for (SObjectField lookupField : recordIdsByLookupField.keySet()) {
+            Set<Id> recordIds = recordIdsByLookupField.get(lookupField);
+            if (recordIds.isEmpty()) {
+                continue;
+            }
+
+            SObjectType sObjectType = new List<Id>(recordIds)[0].getSObjectType();
+            rollups = new Rollups(sObjectType, lookupField);
+
+            List<ServiceDelivery__c> deliveries = attendanceSelector.getDeliveryRecordsToRollup(
+                String.valueOf(lookupField),
+                recordIds
+            );
+
+            rollups.rollupDeliveryRecords(deliveries);
```

Comment:
Should we have a way of passing the lookupfield and record ids to an overloaded rollupDeliveryRecords method so that rollups can ask the servicedeliveryservice for the deliveries.